### DISCUSSION TITLE
Advice on updating admin setting programatically + adding mail server settings and outh without UI

### DISCUSSION THREAD

**Discussion by erikd234** [pocketbase/pocketbase#5503] [Q&A]
There are three things I want to do programatically. I have just figured out how to do one of them.

They are

1. create an admin user through extending with Go 
2. add mail stmp server and api key / login through extending with go
I want to make a function like `configureMail(app, apikey, stmpDomain, username, sendFrom...)`
and then like

3. Add google login / other auth providers through extending with go
and something like `addGoogleAuth(clientsecret, clientId)`

That adds these things automatically if they are not there.

Im trying to avoid using the UI as much as possible since I am constantly messing with sqlite files and erasing these settings when I rerun everything. i just know I will forget to add my google auth before releasing one time and break my logins.

Is there an API for doing the above tasks through extending with go? I see the params table for settings im thinking my answer is there but I coudnt quite get it to play nice. Anything have a quick code sample I could try for solving 2 and 3?

For number 1, adding the admin user, im thinking I just can add the auth record as I would with the regular users collection.

Help much appreciated.|

thanks

**Comment by ganigeorgiev** [comment]
1: See the ["Creating new admin" migrations example](https://pocketbase.io/docs/go-migrations/#creating-new-admin).

2,3: See the ["Initialize default app settings" migrations example](https://pocketbase.io/docs/go-migrations/#initialize-default-application-settings).

If you don't want to execute the code through migrations then you can programmatically update directly the `app.Settings()` object and call `app.Dao().SaveSettings(settings)` (_with the upcoming v0.23 release `Dao` is removed and the save will be just `app.Save(settings)`_).

> i just know I will forget to add my google auth before releasing one time and break my logins.

Note that migrations are usually intended to be committed and part of the repo so I'm not sure whether it will be a good idea to have the Google OAuth2 credentials in a migration.

**Reply by erikd234** [reply]
Thanks, this is great! sorry for not reading the docs careful enough. Thats strike 2 for me.

As for the google outh2 secret and id i have in a .env file so it doesnt get commited I was thinking of having that code in main.go or something similar, as well as a conditional check to see if its already configered to not do it more than one time.

**Reply by erikd234** [reply]
`admin.SetPassword("1234567890")`

I think also admin password should not be migrations in plaintext following the same logic as auth providers. thoughts?

**Reply by ganigeorgiev** [reply]
I'll update the examples with the next release to make it more clear, but it really depends on the desired deployment pattern and how do you plan to distribute your app.

If you really want to create the first admin programmatically one possible approach could be to fill the password from an env variable but there is no guarantee that the env variable will exist on the environment where the migration is running. You can throw an error in case that is what you want to use or alternatively (or as fallback) set a random generated password and print it in the terminal.
For example:
```go
pass := security.RandomString(15)
admin.SetPassword(pass)
println("Admin pass:", pass)
```

**Comment by wkirby** [comment]
I've also been playing around looking for a good way to do this, since I'd really like to have the pocketbase settings committed to source control, rather than managed through toggles in the admin UI (in fact, I'd like the ability to fully disable access to the settings UI in production, but that's a different story).

The docs show how to do this all through migration files, which is _ok_, but only runs once: future changes to app-level settings aren't captured by `--automigrate`, nor are they output with the `./pocketbase migrate collections` command (and there's no `./pocketbase migrate settings`).

Instead, we use the `onBeforeBootstrap` hook to add an `onBeforeServe` hook, which configures all the settings from environment variables. This lets us programmatically enforce app-level configuration in a version controlled way, and any futzing around in the admin panel is overwritten.

Our code looks mostly like this, in a file called `pb_hooks/__bootstrap.pb.js` (double underscore helps ensure this runs before any new hook files that might be added):

```js
/// <reference path="../pb_data/types.d.ts" />

onBeforeBootstrap((e) => {
  e.app.onBeforeServe().add((e) => {
    try {
      const dao = e.app.dao();
      let settings = dao.findSettings();

      // App
      settings.meta.appName = process.env.APP_NAME;
      settings.meta.appUrl = process.env.APP_URL;
      settings.meta.hideControls = true;

      // S3
      settings.s3.enabled = true;
      settings.s3.forcePathStyle = process.env.S3_PATH_STYLE || false;
      settings.s3.endpoint = process.env.S3_URL;
      settings.s3.bucket = process.env.S3_STORAGE_BUCKET;
      settings.s3.region = process.env.S3_REGION;
      settings.s3.accessKey = process.env.S3_ACCESS_KEY;
      settings.s3.secret = process.env.S3_SECRET_KEY;

      settings.backups.cron = "0 0 * * *";
      settings.backups.s3.enabled = true;
      settings.backups.s3.forcePathStyle = process.env.S3_PATH_STYLE || false;
      settings.backups.s3.endpoint = process.env.S3_URL;
      settings.backups.s3.bucket = process.env.S3_BACKUP_BUCKET;
      settings.backups.s3.region = process.env.S3_REGION;
      settings.backups.s3.accessKey = process.env.S3_ACCESS_KEY;
      settings.backups.s3.secret = process.env.S3_SECRET_KEY;

      // SMTP
      settings.smtp.enabled = true;
      settings.smtp.host = process.env.SMTP_HOST;
      settings.smtp.port = process.env.SMTP_PORT;
      settings.smtp.username = process.env.SMTP_USERNAME;
      settings.smtp.password = process.env.SMTP_PASSWORD;

      dao.saveSettings(settings);

      // Create default admin
      if (
        process.env.DEFAULT_ADMIN_EMAIL &&
        process.env.DEFAULT_ADMIN_PASSWORD
      ) {
        const admin = new Admin();
        admin.email = process.env.DEFAULT_ADMIN_EMAIL;
        admin.setPassword(process.env.DEFAULT_ADMIN_PASSWORD);

        dao.saveAdmin(admin);
      }

    } catch (err) {
      console.error(`Error while applying settings configuration: ${err}`);
    }
  });
});
```

**Reply by ganigeorgiev** [reply]
I'm not sure I can recommend doing the above (it doesn't prevent users to set different settings during runtime and the admin creation also seems that will fail on the next app startup).

Being able to set the app settings as part of a migration is mostly helpful (_at least for me_) when you want your application to have custom default app settings on first app startup (_e.g. see the [Presentator example](https://github.com/presentator/presentator/blob/master/migrations/1706354334_default_settings.go)_) but this shouldn't disallow users to adjust the settings for their specific environment.

This is also why the settings are not automigrated because most of the settings differ from one environment to another and it rarely make sense to be tracked and committed in version control. Even when using env variables it is not guaranteed that the variables will be available in the running environment and you may ended up with broken app settings state (_with the upcoming v0.23.0 release and the removal of the `Dao` abstraction, we run validations by default on `app.Save()` which will make it harder to persist invalid settings but still it is not ideal_).

I'm still not sure what to suggest mostly because I don't fully understand what actually is the desired behavior here. I'm willing to invest time supporting whatever users are trying to do but I need first to understand it, so please if anyone want the situation around the settings management to change, try to elaborate on your use case and your ideal/desired flow (references to other apps are also welcomed).

**Reply by jsiedentop** [reply]
I think it's going in the direction of infrastructure as code or automated deployments. At least that would be my use case for this.

I prefer to automate the steps to set up a server. The “playbook” is part of a git repository and can therefore be versioned and possibly also tested automatically. Of course, the secrets do not have to be checked in for this, but it is part of the deployment process to ensure that the secrets are available on the target system (for example as environment variables), before the processes are started.

All the steps are documented in code and therefore it's unlikely that I forget something and break the production environment. Changes such as new features and version updates can be tested without much effort because I can replicate the production environment within seconds.

I would prefer if we could do all the settings via pocketbase cli, but I'm also totally fine with the migration approach.

**Comment by pnmcosta** [comment]
In my app, a go flavoured implementation of PocketBase, I leverage env vars/secrets via an explicit `github.com/caarlos0/env` config that I consume on a migration as such:

```
func init() {
	m.Register(func(db dbx.Builder) error {
		cfg, err := config.New()
		if err != nil {
			log.Printf("%s\n", err)
			return err
		}

		dao := daos.New(db).WithoutHooks()
		settings, _ := dao.FindSettings()
		settings.Meta.AppName = cfg.AppName
		settings.Meta.AppUrl = cfg.Host
		settings.Meta.HideControls = true
		settings.Logs.MaxDays = 7

		settings.FacebookAuth.Enabled = true
		settings.FacebookAuth.ClientId = cfg.Facebook.ApiKey
		settings.FacebookAuth.ClientSecret = cfg.Facebook.ApiSecret

		if cfg.Backups.Data.AccessKey != "" {
			settings.Backups.S3.Enabled = true
			settings.Backups.S3.AccessKey = cfg.Backups.Data.AccessKey
			settings.Backups.S3.Secret = cfg.Backups.Data.Secret
			settings.Backups.S3.Endpoint = cfg.Backups.Data.Endpoint
			settings.Backups.S3.Bucket = cfg.Backups.Data.Bucket
			settings.Backups.S3.Region = cfg.Backups.Data.Region
		}

		if err := dao.SaveSettings(settings); err != nil {
			return err
		}
		return nil
	}, func(db dbx.Builder) error {
		return nil
	})
}
```

Considering that I have a full running CI pipeline deploying to Fly.io (see [this](https://github.com/pocketbase/pocketbase/discussions/5581#discussioncomment-10810007) for more info), whenever I need to rotate vars/keys I update them on my GH env (or fly.io secrets) and create a new migration to push to deploy. It's pretty straightforward, neat and simple IMO.
